### DISCUSSION TITLE
.HandleResult() does not throw an exception when all retries fail

### DISCUSSION THREAD

**Discussion by pehrn** [App-vNext/Polly#2021] [General]
Hi,

I have a TryAndRetry function where I Retry if result != expectedResult, however when it reaches max retries, my tests pass and no exception is thrown from .HandleResult(). Could anyone help?

**Comment by martincostello** [comment]
Please provide some code that demonstrates the issue.

**Comment by peter-csala** [comment]
Based on your really vague description it is hard to provide any valuable suggestions. 
As Martin indicated if you could provide a code sample that would help a lot to resolve your issue.
___
BUT let me try to assist you by making an assumption: You have a policy definition which triggers **only if** the decorated method does not return the expected value.

For the sake of simplicity lets assume the definition looks something like this:
```cs
var retry = Policy
    .HandleResult<int>(result => result != expectedResult)
    .Retry(2);
``` 
then all you need to do is to capture the result of the `Execute` and make an assertion against it.

Something like this
```cs
[Fact]
public void Test1()
{
    //Arrange
    const int expectedResult = 42;
    var retry = Policy
      .HandleResult<int>(result => result != expectedResult)
      .Retry(2);

    //Act
    var actualResult = retry.Execute(() => 41);

    //Assert
    Assert.Equal(expectedResult, actualResult);
}
```

**Reply by pehrn** [reply]
Hello & sorry for the late reply.

Actually, your example is pretty much what I am doing. I wanted to remove the final Assert.That() statement and instead just let the `var actualResult = retry.Execute(() => 41)` throw the exception, however nothing happens, but it probably doesn't work like that - I am still working on my c# skills. I assume Assert statement is necessary, then. Thank you

**Reply by peter-csala** [reply]
Hello Peter,

If your decorated method does not throw exception then a retry policy will not throw either. 
In other cases like timeout, the policy itself throws a `TimeoutRejectedException` if the execution did not finish on time.

The `retry.Execute` will throw exception only if your decorated method throws exception and you run out of retry attempts. This is depicted on this diagram: https://github.com/App-vNext/Polly/wiki/Retry#how-polly-retry-works

So, the `Assert.XYZ` is needed to perform an assessment on the actual and expected values. 
If they mismatch then it will throw an exception which will cause the unit test to fail.
